pageserver/src/tenant/timeline.rs

```diff
@@ -3332,27 +3347,41 @@ impl Timeline {
                 ranges: vec![NON_INHERITED_RANGE, NON_INHERITED_SPARSE_RANGE],
             });
 
-            // Keyspace is fully retrieved, no ancestor timeline, or metadata scan (where we do not look
-            // into ancestor timelines). TODO: is there any other metadata which we want to inherit?
-            if keyspace.total_raw_size() == 0 || timeline.ancestor_timeline.is_none() {
-                break;
+            // Keyspace is fully retrieved
+            if keyspace.is_empty() {
+                break None;
+            }
+
+            // Not fully retrieved but no ancestor timeline.
+            if timeline.ancestor_timeline.is_none() {
+                break Some(keyspace);
             }
 
+            // Now we see if there are keys covered by the image layer but does not exist in the
+            // image layer, which means that the key does not exist.
+
+            // Get the overlapping of the image layer keyspace and the incomplete keyspace.
+            let removed = keyspace.remove_overlapping_with(&image_covered_keyspace);
+            if !removed.is_empty() {
+                break Some(removed);
+            }
```

Comment:
I think this is correct, but l don't think I'll understand this comment in a week's time. Perhaps we can expand on it. Proposal:
```
The block below will stop the vectored search if any of the keys encountered an image layer 
which did not contain a snapshot for said key. Since we have already removed all completed
keys from `keyspace`, we expect there to be no overlap between it and the image covered key
space. If that's not the case, we had at least one key encounter a gap in the image layer
and stop the search as a result of that.
```